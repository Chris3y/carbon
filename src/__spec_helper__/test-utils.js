const isUpper = char => char.toUpperCase() === char;
const humpToDash = (acc, char) => `${acc}${isUpper(char) ? `-${char.toLowerCase()}` : char}`;

export const toCSSCase = (str) => {
  return str.split('').reduce(humpToDash, '');
};

export const assertStyleMatch = (styleSpec, component) => {
  Object.entries(styleSpec).forEach(([attr, value]) => {
    expect(component).toHaveStyleRule(toCSSCase(attr), value);
  });
};
