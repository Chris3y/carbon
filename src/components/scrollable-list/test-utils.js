import React from 'react';
import ScrollableListItem from './scrollable-list-item.component';

const makeArray = n => [...Array(n).keys()];

const listItemReducer = ({ nonSelectables = [], customSelectables = [] }) => {
  // generate jsx for selectable list items, based on indexes in config
  return (acc, item, index) => {
    if (nonSelectables.includes(item)) return [...acc, <div />];
    if (customSelectables.includes(item)) return [...acc, <div isSelectable />];

    return [...acc, <ScrollableListItem id={ index } />];
  };
};

const renderListItems = (opts) => {
  return makeArray(opts.num).reduce(listItemReducer(opts), []);
};


const keyPress = (code) => {
  const ev = new KeyboardEvent('keydown', { which: code });
  document.dispatchEvent(ev);
};

const keyMap = {
  UpArrow: '38',
  DownArrow: '40',
  RightArrow: '39',
  LeftArrow: '37',
  Enter: '13',
  Tab: '9'
};

const repeat = action => (n = 1) => makeArray(n).forEach(() => action());

const keyboard = Object.keys(keyMap).reduce((acc, key) => {
  acc[`press${key}`] = () => repeat(keyPress(keyMap[key]));
  return acc;
}, {});

const listFrom = wrapper => wrapper.find('ul');
const childrenFrom = node => node.children();

const hoverList = wrapper => (item) => {
  childrenFrom(listFrom(wrapper)).at(item).simulate('mouseover');
};

const simulateEvent = eventName => wrapper => wrapper.simulate(eventName);
const click = simulateEvent('click');

const selectedItemOf = wrapper => wrapper.state().selectedItem;
const isUnique = (val, index, self) => self.indexOf(val) === index;
const isSelectableGiven = nonSelectables => i => !nonSelectables.includes(i);

const selectedItemReducer = method => wrapper => (acc, i) => {
  method(wrapper)(i);
  return [...acc, selectedItemOf(wrapper)];
};

const arraysEqual = (arr1, arr2) => arr1.sort().join(',') === arr2.sort().join(',');

const assertCorrectTraversal = method => expect => ({ num, nonSelectables = [] }) => (wrapper) => {
  const array = makeArray(num);
  const validIndexes = array.filter(isSelectableGiven(nonSelectables));

  const indexesThatWereSelected = array
    .reduce(selectedItemReducer(method)(wrapper), [])
    .filter(isUnique);
  expect(arraysEqual(validIndexes, indexesThatWereSelected)).toBeTruthy();
};

const assertKeyboardTraversal = assertCorrectTraversal(() => keyboard.pressDownArrow)(expect);
const assertHoverTraversal = assertCorrectTraversal(wrapper => hoverList(wrapper))(expect);

export {
  hoverList,
  selectedItemOf,
  childrenFrom,
  renderListItems,
  keyboard,
  assertKeyboardTraversal,
  assertHoverTraversal,
  listFrom,
  click
};
