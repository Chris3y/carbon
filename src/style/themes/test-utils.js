import generatePalette from '../palette';
import atOpacity from '../utils/at-opacity';
import baseColors from '../color-config';

const addHex = base => `#${base}`;

export const palette = {
  ...generatePalette(baseColors),
  atOpacity,
  ...Object.keys(baseColors).reduce((acc, col) => {
    acc[col] = addHex(baseColors[col]);
    return acc;
  }, {})
};
const blackWithOpacity = atOpacity('#000000');

export const baseThemeConfig = {
  colors: {
    brand: palette.brilliantGreen,
    white: '#FFFFFF',
    error: palette.errorRed,
    warning: palette.gold,
    success: palette.brilliantGreenShade(20),
    info: palette.productBlueShade(3),
    text: {
      body: blackWithOpacity(0.9),
      disabled: blackWithOpacity(0.55),
      placeholder: blackWithOpacity(0.3)
    }
  }
};


const isObject = (obj) => {
  return typeof obj === 'object' && obj !== null;
};

export const assertIsSubset = (obj, comparison) => {
  if (!isObject(obj)) {
    // no further nesting, assert that values are equal
    expect(obj).toEqual(comparison);
    return;
  }

  const objKeys = Object.keys(obj);
  const comparisonKeys = Object.keys(comparison);

  objKeys.forEach((key) => {
    // assert that keys are present
    expect(comparisonKeys.includes(key)).toBeTruthy();

    // repeat for nested objects
    assertIsSubset(obj[key], comparison[key]);
  });
};
