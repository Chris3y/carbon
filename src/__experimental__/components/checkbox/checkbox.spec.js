import React from 'react';
import { mount } from 'enzyme';
import 'jest-styled-components';
import Checkbox from '.';
import baseTheme from '../../../style/themes/base';
import classicTheme from '../../../style/themes/classic';
import { assertStyleMatch } from '../../../__spec_helper__/test-utils';

import FieldHelpStyle from '../field-help/field-help.style';
import HiddenCheckboxStyle from './hidden-checkbox.style';
import LabelStyle from '../label/label.style';
import StyledHelp from '../help/help.style';

const mountRender = (props) => {
  return mount(<Checkbox { ...props } />);
};

describe('Checkbox', () => {
  describe('base theme', () => {
    it('renders as expected', () => {
      expect(mountRender()).toMatchSnapshot();
    });

    describe('renders the appropriate styles when size=large', () => {
      const wrapper = mountRender({ size: 'large' });

      it('.carbon-checkbox__input', () => {
        assertStyleMatch({
          height: '24px',
          padding: '2px',
          width: '24px'
        }, wrapper.find('.carbon-checkbox__input'));
      });

      it('HiddenCheckboxStyle', () => {
        assertStyleMatch({
          height: '24px',
          padding: '2px',
          width: '24px'
        }, wrapper.find(HiddenCheckboxStyle));
      });

      it('HiddenCheckboxStyle + svg', () => {
        assertStyleMatch({
          height: '24px',
          padding: '2px',
          width: '24px'
        }, wrapper.find(`${HiddenCheckboxStyle} + svg`));
      });

      it('FieldHelpStyle', () => {
        assertStyleMatch({
          marginLeft: '24px',
          paddingLeft: '8px'
        }, wrapper.find(FieldHelpStyle));
      });

      it('LabelStyle', () => {
        assertStyleMatch({
          paddingBottom: '4px',
          paddingLeft: '8px',
          paddingTop: '4px'
        }, wrapper.find(LabelStyle));
      });
    });

    describe('renders the appropriate styles when checked=true', () => {
      it('default', () => {
        const wrapper = mountRender({ checked: true });

        assertStyleMatch({
          fill: baseTheme.colors.primary
        }, wrapper.find('svg path'));
      });

      it('disabled=true', () => {
        const wrapper = mountRender({ checked: true, disabled: true });

        assertStyleMatch({
          fill: baseTheme.disabled.border
        }, wrapper.find('svg .checkbox-check'));
      });
    });

    describe('renders the appropriate styles when disabled=true', () => {
      const wrapper = mountRender({ disabled: true });
      const disabledHoverFocusStyles = {
        outline: 'none',
        cursor: 'not-allowed'
      };

      it('LabelStyle', () => {
        assertStyleMatch({
          color: baseTheme.disabled.disabled
        }, wrapper.find(LabelStyle));
      });

      it('StyledHelp', () => {
        assertStyleMatch({
          color: baseTheme.disabled.disabled
        }, wrapper.find(StyledHelp));
      });

      it('svg', () => {
        assertStyleMatch({
          backgroundColor: baseTheme.disabled.input,
          border: `1px solid ${baseTheme.disabled.border}`
        }, wrapper.find('svg'));
      });

      it('HiddenCheckboxStyle:hover', () => {
        assertStyleMatch(
          disabledHoverFocusStyles,
          wrapper.find(HiddenCheckboxStyle).simulate('hover')
        );
      });

      it('HiddenCheckboxStyle:focus', () => {
        assertStyleMatch(
          disabledHoverFocusStyles,
          wrapper.find(HiddenCheckboxStyle).simulate('focus')
        );
      });

      it('LabelStyle:hover', () => {
        assertStyleMatch(
          disabledHoverFocusStyles,
          wrapper.find(LabelStyle).simulate('hover')
        );
      });

      it('LabelStyle:focus', () => {
        assertStyleMatch(
          disabledHoverFocusStyles,
          wrapper.find(LabelStyle).simulate('focus')
        );
      });

      it('svg:hover', () => {
        assertStyleMatch(
          disabledHoverFocusStyles,
          wrapper.find('svg').simulate('hover')
        );
      });

      it('svg:focus', () => {
        assertStyleMatch(
          disabledHoverFocusStyles,
          wrapper.find('svg').simulate('focus')
        );
      });
    });

    it('renders the appropriate styles when error=true', () => {
      const wrapper = mountRender({ error: true });

      assertStyleMatch({
        border: `1px solid ${baseTheme.colors.error}`
      }, wrapper.find('svg'));
    });

    describe('renders the appropriate styles when fieldHelpInline=true', () => {
      it('default', () => {
        const wrapper = mountRender({ fieldHelpInline: true });

        assertStyleMatch({
          display: 'inline',
          margin: '0',
          width: 'auto'
        }, wrapper.find(FieldHelpStyle));
      });

      it('size=large', () => {
        const wrapper = mountRender({ fieldHelpInline: true, size: 'large' });

        assertStyleMatch({
          paddingTop: '4px',
          paddingBottom: '4px'
        }, wrapper.find(FieldHelpStyle));
      });
    });

    describe('renders the appropriate styles when setting a custom inputWidth', () => {
      describe('default', () => {
        const wrapper = mountRender({ inputWidth: 50 });

        it('.carbon-checkbox__input', () => {
          assertStyleMatch({
            width: '50%'
          }, wrapper.find('.carbon-checkbox__input'));
        });

        it('FieldHelpStyle', () => {
          assertStyleMatch({
            marginLeft: '50%'
          }, wrapper.find(FieldHelpStyle));
        });
      });

      it('reversed', () => {
        const wrapper = mountRender({ inputWidth: 50, reverse: true });

        assertStyleMatch({
          marginRight: '50%'
        }, wrapper.find(FieldHelpStyle));
      });
    });

    it('renders the appropriate styles when setting a custom labelWidth', () => {
      const wrapper = mountRender({ labelWidth: 50 });

      assertStyleMatch({
        width: '50%'
      }, wrapper.find(LabelStyle));
    });

    describe('renders the appropriate styles when reverse=true', () => {
      it('default', () => {
        const wrapper = mountRender({ reverse: true });

        assertStyleMatch({
          marginLeft: '0'
        }, wrapper.find(FieldHelpStyle));
      });

      it('fieldHelpInline=true', () => {
        const wrapper = mountRender({ fieldHelpInline: true, reverse: true });

        assertStyleMatch({
          paddingLeft: '6px'
        }, wrapper.find('.carbon-checkbox__input'));
      });
    });
  });

  describe('Classic theme', () => {
    const opts = { theme: classicTheme };

    it('renders as expected', () => {
      expect(mountRender(opts)).toMatchSnapshot();
    });

    it('renders the appropriate styles when checked=true', () => {
      describe('default', () => {
        expect(mountRender({ checked: true, ...opts })).toMatchSnapshot();
      });

      describe('and disabled=true', () => {
        expect(mountRender({ checked: true, disabled: true, ...opts })).toMatchSnapshot();
      });
    });

    it('renders the appropriate styles when disabled=true', () => {
      expect(mountRender({ disabled: true, ...opts })).toMatchSnapshot();
    });

    it('renders the appropriate styles when fieldHelpInline=true and reverse=true', () => {
      expect(mountRender({ fieldHelpInline: true, reverse: true, ...opts })).toMatchSnapshot();
    });
  });
});
