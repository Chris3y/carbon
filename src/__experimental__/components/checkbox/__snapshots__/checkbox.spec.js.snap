// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox base theme renders as expected 1`] = `
.c4 {
  display: inline-block;
  padding-top: 1px;
  position: relative;
}

.c6 {
  cursor: pointer;
  opacity: 0;
  margin: 0;
  z-index: 999;
}

.c7 .c3 {
  padding-top: 1px;
}

.c7 .c5,
.c7 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c7 .c3,
.c7 .c5,
.c7 svg {
  box-sizing: border-box;
  width: 16px;
}

.c7 .c5:not([disabled]):focus + svg,
.c7 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c7 .c3,
.c7 .c5,
.c7 .c5 + svg {
  height: 24px;
  padding: 2px;
  width: 24px;
}

.c8 .c3 {
  padding-top: 1px;
}

.c8 .c5,
.c8 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c8 .c3,
.c8 .c5,
.c8 svg {
  box-sizing: border-box;
  width: 16px;
}

.c8 .c5:not([disabled]):focus + svg,
.c8 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c8 .c3,
.c8 .c5,
.c8 .c5 + svg {
  height: 24px;
  padding: 2px;
  width: 24px;
}

.c9 .c3 {
  padding-top: 1px;
}

.c9 .c5,
.c9 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c9 .c3,
.c9 .c5,
.c9 svg {
  box-sizing: border-box;
  width: 16px;
}

.c9 .c5:not([disabled]):focus + svg,
.c9 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c9 .c5:hover,
.c9 svg:hover,
.c9 .c19:hover,
.c9 .c5:focus,
.c9 svg:focus,
.c9 .c19:focus {
  outline: none;
  cursor: not-allowed;
}

.c10 .c3 {
  padding-top: 1px;
}

.c10 .c5,
.c10 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c10 .c3,
.c10 .c5,
.c10 svg {
  box-sizing: border-box;
  width: 16px;
}

.c10 .c5:not([disabled]):focus + svg,
.c10 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c10 .c5:hover,
.c10 svg:hover,
.c10 .c19:hover,
.c10 .c5:focus,
.c10 svg:focus,
.c10 .c19:focus {
  outline: none;
  cursor: not-allowed;
}

.c11 .c3 {
  padding-top: 1px;
}

.c11 .c5,
.c11 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c11 .c3,
.c11 .c5,
.c11 svg {
  box-sizing: border-box;
  width: 16px;
}

.c11 .c5:not([disabled]):focus + svg,
.c11 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c11 .c3 {
  width: 50%;
}

.c12 .c3 {
  padding-top: 1px;
}

.c12 .c5,
.c12 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c12 .c3,
.c12 .c5,
.c12 svg {
  box-sizing: border-box;
  width: 16px;
}

.c12 .c5:not([disabled]):focus + svg,
.c12 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c13 .c3 {
  padding-top: 1px;
}

.c13 .c5,
.c13 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c13 .c3,
.c13 .c5,
.c13 svg {
  box-sizing: border-box;
  width: 16px;
}

.c13 .c5:not([disabled]):focus + svg,
.c13 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c13 .c3 {
  padding-left: 6px;
}

.c14 .c3 {
  padding-top: 1px;
}

.c14 .c5,
.c14 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c14 .c3,
.c14 .c5,
.c14 svg {
  box-sizing: border-box;
  width: 16px;
}

.c14 .c5:not([disabled]):focus + svg,
.c14 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c14 .c3 {
  padding: 1px 0 0 0;
}

.c14 .c5,
.c14 svg {
  padding: 1px;
}

.c14 .c3,
.c14 .c5,
.c14 .c5 + svg {
  height: 15px;
  width: 15px;
}

.c14 .c5:not([disabled]):focus + svg,
.c14 .c5:not([disabled]):hover + svg {
  border: 1px solid #1963f6;
  outline: none;
}

.c15 .c3 {
  padding-top: 1px;
}

.c15 .c5,
.c15 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c15 .c3,
.c15 .c5,
.c15 svg {
  box-sizing: border-box;
  width: 16px;
}

.c15 .c5:not([disabled]):focus + svg,
.c15 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c15 .c5:hover,
.c15 svg:hover,
.c15 .c19:hover,
.c15 .c5:focus,
.c15 svg:focus,
.c15 .c19:focus {
  outline: none;
  cursor: not-allowed;
}

.c15 .c3 {
  padding: 1px 0 0 0;
}

.c15 .c5,
.c15 svg {
  padding: 1px;
}

.c15 .c3,
.c15 .c5,
.c15 .c5 + svg {
  height: 15px;
  width: 15px;
}

.c15 .c5:not([disabled]):focus + svg,
.c15 .c5:not([disabled]):hover + svg {
  border: 1px solid #1963f6;
  outline: none;
}

.c16 .c3 {
  padding-top: 1px;
}

.c16 .c5,
.c16 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c16 .c3,
.c16 .c5,
.c16 svg {
  box-sizing: border-box;
  width: 16px;
}

.c16 .c5:not([disabled]):focus + svg,
.c16 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c16 .c5:hover,
.c16 svg:hover,
.c16 .c19:hover,
.c16 .c5:focus,
.c16 svg:focus,
.c16 .c19:focus {
  outline: none;
  cursor: not-allowed;
}

.c16 .c3 {
  padding: 1px 0 0 0;
}

.c16 .c5,
.c16 svg {
  padding: 1px;
}

.c16 .c3,
.c16 .c5,
.c16 .c5 + svg {
  height: 15px;
  width: 15px;
}

.c16 .c5:not([disabled]):focus + svg,
.c16 .c5:not([disabled]):hover + svg {
  border: 1px solid #1963f6;
  outline: none;
}

.c17 .c3 {
  padding-top: 1px;
}

.c17 .c5,
.c17 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c17 .c3,
.c17 .c5,
.c17 svg {
  box-sizing: border-box;
  width: 16px;
}

.c17 .c5:not([disabled]):focus + svg,
.c17 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c17 .c3 {
  padding-left: 6px;
}

.c17 .c3 {
  padding: 1px 0 0 0;
}

.c17 .c5,
.c17 svg {
  padding: 1px;
}

.c17 .c3,
.c17 .c5,
.c17 .c5 + svg {
  height: 15px;
  width: 15px;
}

.c17 .c5:not([disabled]):focus + svg,
.c17 .c5:not([disabled]):hover + svg {
  border: 1px solid #1963f6;
  outline: none;
}

.c17 .c3 {
  padding-left: 6px;
}

.c0 {
  padding-top: 8px;
}

.c0 .c3 {
  padding-top: 1px;
}

.c0 svg {
  background-color: #FFFFFF;
  border: solid 1px #668491;
}

.c0 .c5,
.c0 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c0 .c3,
.c0 .c5,
.c0 svg {
  box-sizing: border-box;
  width: 16px;
}

.c0 .c5:not([disabled]):focus + svg,
.c0 .c5:not([disabled]):hover + svg {
  outline: solid 3px #FFDA7F;
}

.c0 .c19 {
  padding: 0 6px;
  text-align: left;
  width: auto;
}

.c0 .c19 .c20 {
  color: rgba(0,0,0,0.65);
  vertical-align: bottom;
}

.c0 .c19 .c20:hover,
.c0 .c19 .c20:focus {
  color: rgba(0,0,0,0.9);
}

.c0 .c18 {
  margin-left: 16px;
  margin-top: 0;
  padding-left: 6px;
}

.c2 {
  display: block;
}

.c2 + .c1 {
  margin-top: 16px;
}

<div
  className="c0"
  data-component="checkbox"
>
  <div
    className="c1 c2"
  >
    <div
      className="c3 c4"
    >
      <input
        className="c5 c6"
        id="guid-12345"
        role="checkbox"
        type="checkbox"
      />
      <svg
        height="10"
        viewBox="0 0 12 10"
        width="12"
      >
        <path
          d="M.237 6.477A.752.752 0 0 1 .155 5.47l.851-1.092a.63.63 0 0 1 .934-.088l2.697 1.964, 4.674-6a.63.63 0 0 1 .933-.088l1.015.917c.28.254.317.703.081 1.005L6.353 8.492a.725.725, 0 0 1-.095.16l-.85 1.093a.637.637 0 0 1-.626.244.638.638 0 0 1-.335-.16L.237 6.476z"
          fill="#FFFFFF"
          fillRule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;
